src/test/firestore/indexes.spec.ts

```diff
@@ -0,0 +1,221 @@
+      fieldPath: "foo",
+      indexes: [
+        { order: "ASCENDING", scope: "COLLECTION" },
+        { arrayConfig: "CONTAINS", scope: "COLLECTION" },
+      ],
+    },
+  ],
+};
+
+describe("IndexValidation", () => {
+  it("should accept a valid v1beta2 index spec", () => {
+    idx.validateSpec(VALID_SPEC);
+  });
+
+  it("should not change a valid v1beta2 index spec after upgrade", () => {
+    const upgraded = idx.upgradeOldSpec(VALID_SPEC);
+    expect(upgraded).to.eql(VALID_SPEC);
+  });
+
+  it("should accept a valid v1beta1 index spec after upgrade", () => {
+    idx.validateSpec(
+      idx.upgradeOldSpec({
+        indexes: [
+          {
+            collectionId: "collection",
+            fields: [
+              { fieldPath: "foo", mode: "ASCENDING" },
+              { fieldPath: "bar", mode: "DESCENDING" },
+              { fieldPath: "baz", mode: "ARRAY_CONTAINS" },
+            ],
+          },
+        ],
+      })
+    );
+  });
+
+  it("should reject an incomplete index spec", () => {
+    expect(() => {
+      idx.validateSpec({
+        indexes: [
+          {
+            collectionGroup: "collection",
+            fields: [
+              { fieldPath: "foo", order: "ASCENDING" },
+              { fieldPath: "bar", order: "DESCENDING" },
+            ],
+          },
+        ],
+      });
+    }).to.throw();
+  });
+
+  it("should reject an overspecified index spec", () => {
+    expect(() => {
+      idx.validateSpec({
+        indexes: [
+          {
+            collectionGroup: "collection",
+            queryScope: "COLLECTION",
+            fields: [
+              { fieldPath: "foo", order: "ASCENDING", arrayConfig: "CONTAINES" },
+              { fieldPath: "bar", order: "DESCENDING" },
+            ],
+          },
+        ],
+      });
+    }).to.throw();
+  });
+});
+
+describe("IndexNameParsing", () => {
+  it("should parse an index name correctly", () => {
+    const name =
+      "/projects/myproject/databases/(default)/collectionGroups/collection/indexes/abc123/";
+    expect(idx.parseIndexName(name)).to.eql({
+      projectId: "myproject",
+      collectionGroupId: "collection",
+      indexId: "abc123",
+    });
+  });
+
+  it("should parse a field name correctly", () => {
+    const name =
+      "/projects/myproject/databases/(default)/collectionGroups/collection/fields/abc123/";
+    expect(idx.parseFieldName(name)).to.eql({
+      projectId: "myproject",
+      collectionGroupId: "collection",
+      fieldPath: "abc123",
+    });
+  });
+});
+
+describe("IndexSpecMatching", () => {
+  it("should identify a positive index spec match", () => {
+    const apiIndex = {
+      name: "/projects/myproject/databases/(default)/collectionGroups/collection/indexes/abc123",
+      queryScope: "COLLECTION",
+      fields: [
+        { fieldPath: "foo", order: "ASCENDING" },
+        { fieldPath: "bar", arrayConfig: "CONTAINS" },
+      ],
+      state: API.State.READY,
+    } as API.Index;
```

Comment:
this is definitely a hole in my TS knowledge - is this equivalent? (with no `as API.Index`)

```typescript
const apiIndex: API.Index = {...};
```

Reply:
Yeah I am not sure what's going on here ... if I do what you suggested it wants me to define all undefined fields as well.

Reply:
Oh wait I figured it out.